sockets.js

```diff
@@ -549,10 +549,8 @@ if (cluster.isMaster) {
 			process.send(`<${socketid}\n${message}`);
 		});
 
-		socket.once('close', () => {
-			process.send(`!${socketid}`);
-			sockets.delete(socketid);
-			channels.forEach(channel => channel.delete(socketid));
+		socket.on('close', () => {
+			if (sockets.has(socketid)) process.send(`!${socketid}`);
```

Comment:
This will get `!socketid` back?

Reply:
This checks whether or not another close event has removed `socketid` from `sockets`. `socketid` never changes value

Reply:
`Sockets.socketDisconnect` is the only way sockets can be cleaned up in this, since regardless of why a socket was closed, `Connection` instances still need to be cleaned up beforehand. The only case it doesn't is when the connection closes immediately after beginning to connect, since the socket hasn't been added to `sockets` yet, but that's taken care of by `sweepSocketInterval`

Reply:
I mean, sending `!socketid` to the main process will make the main process send `!socketid` back?

Reply:
Ohh. Yeah, it will so long as a connection exists for the socket. That leaves the risk of a connection closing while its list of IPs gets checked for any proxies, making it possible that `!<socketid>` could get sent before `*<socketid>,<ip>,<protocol>` though. Maybe it'd be better to just make it always send `!<socketid>`  for now, since I can't think of a good way to prevent sending it redundantly offhand.

Reply:
Why not just close it here, and just ignore any `!<socketid>` that the main process sends?

Reply:
The parent process should be aware when sockets are going to close before the fact so it doesn't try to send messages to them after they've closed as much as it does now. Closing it from the message handler instead prevents this from happening with sockets the server doesn't close itself.

However, the sockets that would send those messages out of order happen to be the same ones `sweepSocketInterval` close periodically. Closing only those specific sockets from here would both fix this and remove the need for the timer to exist, depending on whether or not sockets using XHR Streaming as their transport really need to be closed every 10 minutes

Reply:
Why is sending messages to them after they've closed a problem? It should not actually be a problem...

Reply:
It's only a problem with sockets that end up sending their connect message after they have disconnected. With others, is it really a bad thing to prevent them from wasting IO resources on messages that mean nothing in the long run?

Reply:
...none of those are really flaws, but I suppose this logic _is_ simpler.